File: clients/python/linaldb/dataset.py
"""`/delivery` dataset export — see clients/CONTRACT.md §2. This is the
checkpoint-2 priority workload (saved-dataset Parquet export), chosen
over ad-hoc /execute query-result ergonomics in the design discussion
that started this plan.
"""

from __future__ import annotations

import io
import json

from .errors import LinalError


def _vector_or_matrix_columns(schema: dict) -> dict:
    """Map column name -> declared value_type for every Vector/Matrix
    column in a `schema.json` document (contract §2's "authoritative
    column typing" — used to detect the legacy JSON-string fallback
    encoding, since that can't be told apart from a real string column
    by looking at the Parquet file alone).
    """
    result = {}
    for col in schema["columns"]:
        vt = col["value_type"]
        if isinstance(vt, dict) and ("Vector" in vt or "Matrix" in vt):
            result[col["name"]] = vt
    return result


def _unwrap_json_fallback_cell(raw: str | None):
    if raw is None:
        return None
    parsed = json.loads(raw)
    # Same tagged shape as a live /execute Value cell (contract §3) --
    # {"Vector": [...]} / {"Matrix": [[...]]} -- reuse the same unwrap.
    from .wire import unwrap_value

    return unwrap_value(parsed)


class Dataset:
    def __init__(self, client, name: str):
        self._client = client
        self.name = name

    def _delivery_url(self, path: str) -> str:
        return f"{self._client.url}/delivery/datasets/{self.name}/{path}"

    def _get_json(self, path: str) -> dict:
        resp = self._client._session.get(self._delivery_url(path), timeout=30)
        if resp.status_code != 200:
            raise LinalError(
                f"GET /delivery/datasets/{self.name}/{path} failed "
                f"(HTTP {resp.status_code}): {resp.text}"
            )
        return resp.json()

    def manifest(self) -> dict:
        return self._get_json("manifest.json")

    def schema(self) -> dict:
        return self._get_json("schema.json")

    def stats(self) -> dict:
        return self._get_json("stats.json")

    def to_arrow(self):
        """Read `data.parquet` as a `pyarrow.Table`, transparently
        unwrapping any column that landed in the legacy JSON-string
        fallback encoding (contract §2) back into a real nested-list
        column — the caller never sees the raw `{"Vector": [...]}` text.
        """
        import pyarrow as pa
        import pyarrow.parquet as pq

        resp = self._client._session.get(self._delivery_url("data.parquet"), timeout=60)
        if resp.status_code != 200:
            raise LinalError(
                f"GET /delivery/datasets/{self.name}/data.parquet failed "
                f"(HTTP {resp.status_code}): {resp.text}"
            )
        table = pq.read_table(io.BytesIO(resp.content))

        vector_or_matrix = _vector_or_matrix_columns(self.schema())
        for col_name in vector_or_matrix:
            if col_name not in table.column_names:
                continue
            idx = table.column_names.index(col_name)
            field_type = table.schema.field(idx).type
            if pa.types.is_string(field_type) or pa.types.is_large_string(field_type):
                # Legacy fallback encoding -- unwrap each cell's JSON text.
                # Note: pa.array() infers float64 for a plain Python list of
                # floats, so this path widens f32 -> f64 (the values
                # themselves are unaffected -- the JSON text already fixed
                # their precision at serde_json::to_string time on the
                # server). The native FixedSizeList path is genuinely
                # float32, matching the engine's f32-only Value::Float type.
                raw_values = table.column(idx).to_pylist()
                unwrapped = [_unwrap_json_fallback_cell(v) for v in raw_values]
                table = table.set_column(idx, col_name, pa.array(unwrapped))
            # Otherwise it's already a native FixedSizeList column
            # (pyarrow/pandas read it correctly with no help needed).

        return table

    def to_pandas(self):
        """`.to_arrow().to_pandas()`. Requires the `pandas` extra."""
        try:
            import pandas  # noqa: F401
        except ImportError as e:
            raise ImportError(
                "Dataset.to_pandas() requires the `pandas` extra: "
                'pip install "linaldb[pandas]"'
            ) from e
        return self.to_arrow().to_pandas()

File: clients/r/R/dataset.R
# /delivery dataset export -- see clients/CONTRACT.md Sec.2. Mirrors
# clients/python/linaldb/dataset.py; this is checkpoint 4 of
# PYTHON_R_INTEROP_PLAN.md, the R counterpart of checkpoint 2's Python
# saved-dataset Parquet export.

#' A handle to a saved dataset's `/delivery/*` export
#'
#' @param conn A `linal_connection` from `linal_connect()`.
#' @param name Dataset name.
#' @return A `linal_dataset` object, passed to `linal_dataset_schema()`/
#'   `linal_dataset_manifest()`/`linal_dataset_stats()`/
#'   `linal_dataset_to_arrow()`/`linal_dataset_read()`.
#' @export
linal_dataset <- function(conn, name) {
  structure(list(conn = conn, name = name), class = "linal_dataset")
}

.linal_delivery_url <- function(ds, path) {
  sprintf("%s/delivery/datasets/%s/%s", ds$conn$url, ds$name, path)
}

#' /delivery resolves the same per-database subdirectory /execute does via
#' this header (contract Sec.2, engine v0.1.73) -- without it, a
#' connection configured for a non-default database would silently hit
#' the *default* database's copy of a same-named dataset instead (or 404
#' if there isn't one), not the one this connection actually points at.
#' Found while building a real example against a non-default database,
#' not caught by checkpoint 4's tests, which only ever used the default
#' database. Mirrors the equivalent fix in the Python client's dataset.py.
#' @keywords internal
#' @noRd
.linal_delivery_request <- function(ds, path, timeout_secs) {
  req <- httr2::request(.linal_delivery_url(ds, path))
  req <- httr2::req_timeout(req, timeout_secs)
  if (!is.null(ds$conn$database)) {
    req <- httr2::req_headers(req, `X-Linal-Database` = ds$conn$database)
  }
  req
}

.linal_delivery_get_json <- function(ds, path) {
  req <- .linal_delivery_request(ds, path, 30)
  resp <- httr2::req_perform(httr2::req_error(req, is_error = function(resp) FALSE))
  if (httr2::resp_status(resp) != 200) {
    stop(linal_error(sprintf(
      "GET /delivery/datasets/%s/%s failed (HTTP %d): %s",
      ds$name, path, httr2::resp_status(resp), httr2::resp_body_string(resp)
    )))
  }
  httr2::resp_body_json(resp, simplifyVector = FALSE)
}

#' @rdname linal_dataset
#' @param ds A `linal_dataset` from `linal_dataset()`.
#' @export
linal_dataset_manifest <- function(ds) {
  .linal_delivery_get_json(ds, "manifest.json")
}

#' @rdname linal_dataset
#' @export
linal_dataset_schema <- function(ds) {
  .linal_delivery_get_json(ds, "schema.json")
}

#' @rdname linal_dataset
#' @export
linal_dataset_stats <- function(ds) {
  .linal_delivery_get_json(ds, "stats.json")
}

#' Column names whose declared `value_type` (per `schema.json`, contract
#' Sec.2's authoritative source for logical column typing) is Vector or
#' Matrix.
#' @keywords internal
#' @noRd
.vector_or_matrix_columns <- function(schema) {
  names_out <- character(0)
  for (col in schema$columns) {
    vt <- col$value_type
    if (is.list(vt) && (!is.null(vt$Vector) || !is.null(vt$Matrix))) {
      names_out <- c(names_out, col$name)
    }
  }
  names_out
}

#' Unwrap one legacy-fallback-encoded cell (a JSON string like
#' `{"Vector": [1.0,2.0,3.0]}`, contract Sec.2) into a plain R value.
#' @keywords internal
#' @noRd
.unwrap_json_fallback_cell <- function(raw) {
  if (is.na(raw)) {
    return(NA)
  }
  unwrap_value(jsonlite::fromJSON(raw, simplifyVector = FALSE))
}

#' Fetch `data.parquet` and return it as a `data.frame`, transparently
#' unwrapping any column that landed in the legacy JSON-string fallback
#' encoding (contract Sec.2) back into a real list-column -- the caller
#' never sees the raw tagged-JSON text (e.g. Vector wire-tagged as
#' \code{list(Vector = c(...))}).
#'
#' @rdname linal_dataset
#' @export
linal_dataset_read <- function(ds) {
  req <- .linal_delivery_request(ds, "data.parquet", 60)
  resp <- httr2::req_perform(httr2::req_error(req, is_error = function(resp) FALSE))
  if (httr2::resp_status(resp) != 200) {
    stop(linal_error(sprintf(
      "GET /delivery/datasets/%s/data.parquet failed (HTTP %d): %s",
      ds$name, httr2::resp_status(resp), httr2::resp_body_string(resp)
    )))
  }
  raw_bytes <- httr2::resp_body_raw(resp)

  table <- arrow::read_parquet(raw_bytes, as_data_frame = FALSE)
  df <- as.data.frame(table)

  vector_or_matrix <- .vector_or_matrix_columns(linal_dataset_schema(ds))
  for (col_name in vector_or_matrix) {
    if (!(col_name %in% names(df))) {
      next
    }
    field_idx <- match(col_name, table$schema$names) - 1L
    field_type <- table$schema$field(field_idx)$type
    if (inherits(field_type, "Utf8") || inherits(field_type, "LargeUtf8")) {
      # Legacy fallback encoding -- unwrap each cell's JSON text.
      df[[col_name]] <- lapply(df[[col_name]], .unwrap_json_fallback_cell)
    }
    # Otherwise it's already a native FixedSizeList column, which
    # as.data.frame() on the Table already turned into a correct
    # numeric list-column with no help needed.
  }

  df
}

#' `arrow::arrow_table(linal_dataset_read(ds))` -- the fixed-up
#' `data.frame` re-encoded as a real Arrow Table, for callers who want to
#' keep working in Arrow rather than convert to base R types. Rebuilding
#' from the already-unwrapped data.frame (rather than patching the raw
#' Arrow Table's columns directly) keeps a single source of truth with
#' `linal_dataset_read()` instead of two parallel unwrapping
#' implementations that could drift apart.
#'
#' @rdname linal_dataset
#' @export
linal_dataset_to_arrow <- function(ds) {
  arrow::arrow_table(linal_dataset_read(ds))
}
